lib/index.ts

```diff
@@ -0,0 +1,405 @@
+
+export interface RankingsRequest {
+  limit?: RankingsLimit;
+  sortBy: RankingsSortBy;
+  sortMethod: RankingsSortMethod;
+}
+
+export interface RankingsData {
+  id: string;
+  name: string;
+  monthly_votes: number;
+  monthly_votes_rank: number;
+  server_count: number;
+  server_count_rank: number;
+  total_votes: number;
+  total_votes_rank: number;
+  shard_count: number;
+  shard_count_rank: number;
+  monthly_votes_rank_change: number;
+  server_count_rank_change: number;
+  total_votes_rank_change: number;
+  shard_count_rank_change: number;
+}
+
+export interface RankingsResponse {
+  totalBotCount: number;
+  data: RankingsData[];
+}
+
+/**
+ * Custom error classes for better error handling
+ */
+export class TopStatsError extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = "TopStatsError";
+    Object.setPrototypeOf(this, TopStatsError.prototype);
+  }
+}
+
+export class RateLimitError extends TopStatsError {
+  constructor(
+    message: string,
+    public expiresIn: string,
+  ) {
+    super(message);
+    this.name = "RateLimitError";
+    Object.setPrototypeOf(this, RateLimitError.prototype);
+  }
+}
+
+export interface ClientOptions {
+  /** API Token for authentication */
+  token: string;
+  /** Optional custom base URL */
+  baseUrl?: string;
+}
+
+/**
+ * Main client class for interacting with the TopStats API
+ * @example
+ * ```typescript
+ * const client = new Client({ token: 'your-token' });
+ * // or
+ * const client = new Client('your-token');
+ *
+ * // Get bot info
+ * const bot = await client.getBot('123456789');
+ * ```
+ */
+export class Client {
+  private readonly token: string;
+  private readonly BASE_URL: string;
+
+  /**
+   * Creates a new TopStats API client
+   * @param options - Token string or client options object
+   */
+  constructor(options: string | ClientOptions) {
+    if (typeof options === "string") {
+      this.token = options;
+      this.BASE_URL = "https://api.topstats.gg";
+    } else {
+      this.token = options.token;
+      this.BASE_URL = options.baseUrl ?? "https://api.topstats.gg";
+    }
+  }
+
+  /**
+   * Validates a Discord bot ID format
+   * @throws {TopStatsError} If ID format is invalid
+   */
+  private validateBotId(botId: string): void {
+    if (!/^\d{17,19}$/.test(botId)) {
+      throw new TopStatsError("Invalid Discord bot ID format");
+    }
+  }
+
+  /**
+   * Makes an HTTP request to the API
+   * @throws {RateLimitError} When rate limited
+   * @throws {TopStatsError} For other API errors
+   */
+  private async _request<T>(
+    method: string,
+    path: string,
+    query?: Record<string, any>,
+  ): Promise<T> {
+    try {
+      const url = new URL(`${this.BASE_URL}${path}`);
+
+      if (query && method === "GET") {
+        Object.entries(query).forEach(([key, value]) => {
+          if (value !== undefined) {
+            url.searchParams.append(key, String(value));
+          }
+        });
+      }
```

Comment:
This can just be
`url.search = new URLSearchParams(query)`

Reply:
I think in terms of longevity my solution is more appropriate here, as using URLSearchParams will not handle multiple values with the same key, and although any key with an undefined/null value will get stripped from my approach, the API itself should handle missing parameters.

e.g.
```ts
const query = {
  name: "John",
  age: undefined,
  tags: ["a", "b"]
};

// my approach
// ?name=John&tags=a&tags=b

// URLSearchParams
// ?name=John&age=undefined&tags=b
```

Not entirely sure on this though I would love to dicuss further.